ark/utils/spatial_analysis_utils.py

```diff
@@ -246,33 +244,54 @@ def compute_close_cell_num_random(marker_nums, dist_mat, dist_lim, bootstrap_num
             Large matrix of random positive marker counts for every permutation in the bootstrap
     """
 
-    # Get marker counts
-    marker_count = len(marker_nums)
-
-    # Create close_num_rand
-    close_num_rand = np.zeros((
-        marker_count, marker_count, bootstrap_num), dtype=np.uint16)
-
     # Generate binarized distance matrix
-    dist_mat_bin = (dist_mat.values < dist_lim).astype(np.uint16)
-
-    # static choice array
-    choice_ar = list(range(dist_mat.shape[0]))
-
-    for j, m1n in enumerate(marker_nums):
-        for k, m2n in enumerate(marker_nums[j:], j):
-            for r in range(bootstrap_num):
-                # Select same amount of random cell labels as positive ones in close_num
-                marker1_labels_rand = np.random.choice(a=choice_ar, size=m1n, replace=False)
-                marker2_labels_rand = np.random.choice(a=choice_ar, size=m2n, replace=False)
-
-                # Record the number of interactions and store in close_num_rand in the index
-                # corresponding to both markers, for every permutation
-                close_num_rand[j, k, r] = \
-                    np.sum(dist_mat_bin[np.ix_(marker1_labels_rand, marker2_labels_rand)])
-
-            # System should be symetric
-            close_num_rand[k, j, :] = close_num_rand[j, k, :]
+    dist_mat_bin = ((dist_mat.values < dist_lim) & (dist_mat.values > 0)).astype(np.uint16)
+
+    # assures that marker counts don't exceed number of cells
+    for mn in marker_nums:
+        if mn >= dist_mat_bin.shape[0]:
+            raise ValueError('Marker number count can not be greater than number of cells...')
+
+    # creates list-of-lists sparse matrix-representation of binarized distance matrix
+    rows, cols = np.nonzero(dist_mat_bin)
+    cols_in_row = [[] for i in range(dist_mat_bin.shape[0])]
+    for pair_idx, row in enumerate(rows):
+        cols_in_row[row].append(cols[pair_idx])
+
+    # flattens list-of-list representation into 1D array and stores the index keys
+    row_indicies = [0]
+    cols_in_row_flat = []
+    for cell_indx, row in enumerate(range(dist_mat_bin.shape[0])):
+        row_indicies.append(len(cols_in_row[row]) + row_indicies[cell_indx])
+        cols_in_row_flat.extend(cols_in_row[row])
```

Comment:
After `cols_in_row` is created it's then immediately used to create `cols_in_row_flat`. Any chance this can be done in one loop and not two consecutive ones?

Reply:
yep yep yep, good catch